workload/schemachange: re-enable DROP SCHEMA

spilchen opened the pull request:
The schemachange workload aims to predict errors when dropping schemas. However, the query in `schemaContainsTypesWithCrossSchemaReferences` missed handling two specific cases:

1. Cross-Schema Table Reference: If a table in another schema references a type within the schema being dropped, the query failed to identify the table. As a result, the system incorrectly assumed that the drop would succeed, even though it would actually be blocked.

2. Cross-Schema UDF Reference: On the other hand, if a UDF in another schema depends on a type in the schema being dropped, the function is correctly dropped during cascade. The test didn’t handle this situation properly as it thought that the drop should be blocked.

Instead of fixing the error-prediction logic for DROP SCHEMA, I’ve removed the query altogether. I will now flag a potential error when dropping a type that is referenced in another schema to cover the first scenario. This also re-enables the DROP SCHEMA operation in the workload.

Closes: #128345
Closes: #127977
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128705)
<!-- Reviewable:end -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128705#-:-O47GGaIDlOz18hHfTj8:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @srosenberg)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128705)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @spilchen, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 3930 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128705#-O47QHNK06xpNVoVVnxa:-O47QHNK06xpNVoVVnxb:bdte017) ([raw file](https://github.com/cockroachdb/cockroach/blob/f745b4377833d2d6b2796de6625cb91b9b19c597/pkg/workload/schemachange/operation_generator.go#L3930)):*
> ```Go
> 	// referenced in another schema. This can be removed once issue #51480 is
> 	// addressed.
> 	stmt.potentialExecErrors.add(pgcode.FeatureNotSupported)
> ```

one additional thought: is there a way we can avoid universally adding this code? it's a pretty broad one. maybe we only add it if the schema has any user-defined types? this is just a suggestion in case you think it would be easy to add that.

(if we had a way to match an error more specifically than the code so it specifically caught the cross-schema references error, i think that would also be fine. but i don't think it's easy to do that.)


<!-- Sent from Reviewable.io -->

spilchen committed:
workload/schemachanger: re-enable DROP SCHEMA

The schemachanger workload aims to predict errors when dropping schemas.
However, the query in schemaContainsTypesWithCrossSchemaReferences
missed handling two specific cases:

1. Cross-Schema Table Reference: If a table in another schema references a
type within the schema being dropped, the query failed to identify the
table. As a result, the system incorrectly assumed that the drop would
succeed, even though it would actually be blocked.

2. Cross-Schema UDF Reference: On the other hand, if a user-defined
function (UDF) in another schema depends on a type in the schema being
dropped, the function is correctly cascaded and dropped. The test didn’t
handle this situation properly.

Instead of fixing the error-prediction logic for DROP SCHEMA, I’ve
removed most of the query altogether. I will now only flag a potential
error when dropping a schema that has a user defined type. This is in
case the type is referenced in another schema to cover the first
scenario. This also re-enables the DROP SCHEMA operation in the workload.

Closes: #128345
Closes: #127977
Release note: None

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128705)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rafiss, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 3930 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128705#-O47QHNK06xpNVoVVnxa:-O4Avha4Be0vd26vr_9N:bcwvk4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f745b4377833d2d6b2796de6625cb91b9b19c597/pkg/workload/schemachange/operation_generator.go#L3930)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

one additional thought: is there a way we can avoid universally adding this code? it's a pretty broad one. maybe we only add it if the schema has any user-defined types? this is just a suggestion in case you think it would be easy to add that.

(if we had a way to match an error more specifically than the code so it specifically caught the cross-schema references error, i think that would also be fine. but i don't think it's easy to do that.)
</blockquote></details>

That's a great point, and I’ve incorporated your suggestion. I also included elements from the previous query we used to detect cross-schema dependencies. Since we’re now simply checking whether a user-defined type exists in the schema, the query is much simpler and should be easier to maintain moving forward.


<!-- Sent from Reviewable.io -->

spilchen commented:
tftr!

bors r=rafiss
